### PR TITLE
Support Tuples for Layer Attribution

### PR DESCRIPTION
**Pull request by vivekmig**
This PR adds support for layer inputs / outputs being tuples for all layer attribution algorithms. This also makes corresponding updates to documentation for each modified algorithm.

For all layer attribution algorithms, if the layer input / output is a single tensor, a tensor of attributions is still returned (maintaining consistency with the existing behavior), and for multiple tensors, a tuple is returned. NoiseTunnel was currently using inputs to decide whether the output should be a tuple, that behavior has been modified to be consistent with the described behavior for Layer Attribution.

This PR also contains a bug fix for targets in LayerDeepLift, which was not using the given targets.

### CODE REVIEW

**Comment by vivekmig** [comment]
Updated to handle tuples in LayerIntegratedGradients and fixed condition in NoiseTunnel with new logic for layer attribution tuples.

File: captum/attr/_utils/gradient.py [hunk 1]
```diff
@@ -367,23 +371,32 @@ def compute_layer_gradients_and_eval(
         # If only one key exists (standard model), key list simply has one element.
         key_list = _sort_key_list(list(saved_layer.keys()), device_ids)
         all_outputs = _reduce_list([saved_layer[device_id] for device_id in key_list])
-        grad_inputs = tuple(saved_layer[device_id] for device_id in key_list)
+        num_tensors = len(saved_layer[next(iter(saved_layer))])
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_utils/gradient.py]: Good point, sure, will do!

File: tests/attr/helpers/utils.py [hunk 2]
```diff
@@ -19,13 +19,28 @@ def assertArraysAlmostEqual(inputArr, refArr, delta=0.05):
         )
 
 
-def assertTensorAlmostEqual(test, tensor, expected, delta=0.0001):
+def assertTensorAlmostEqual(test, tensor, expected, delta=0.0001, mode="sum"):
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply tests/attr/helpers/utils.py]: Yes, that's better naming, will update, thanks!

File: tests/attr/neuron/test_neuron_gradient.py [hunk 3]
```diff
@@ -116,14 +116,14 @@ def _gradient_input_test_assert(
             )
 
     def _gradient_matching_test_assert(self, model, output_layer, test_input):
-        out = _forward_layer_eval(model, test_input, output_layer)
+        out = _forward_layer_eval(model, test_input, output_layer)[0][0]
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply tests/attr/neuron/test_neuron_gradient.py]: Yes, the second one is also necessary since it chooses the tensor from the tuple, will add a comment to make it clearer.

File: captum/attr/_core/noise_tunnel.py [hunk 4]
```diff
@@ -283,6 +283,10 @@ def compute_expected_attribution_and_sq(attribution):
 
         if self.is_delta_supported and return_convergence_delta:
             attributions, delta = attributions
+        if isinstance(self.attribution_method, LayerAttribution):
+            # For layer attribution, we want to wrap the attributions in a tuple
+            # only when the returned attributions are in a tuple.
+            is_inputs_tuple = isinstance(attributions, tuple)
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/noise_tunnel.py]: I think the only issue with doing this in general is that we internally convert to a tuple to add noise and pass the tuple as inputs, so the returned attributions currently will always be a tuple for non-layer cases. We would need to unpack inputs based on is_inputs_tuple before passing to the method, I can switch to that instead?

File: captum/attr/_utils/batching.py [hunk 5]
```diff
@@ -25,13 +25,16 @@ def _reduce_list(val_list, red_func=torch.cat):
...
     function to corresponding elements of each tuple in the list, and returns
     tuple of reduction function outputs with length matching the length of tuple
     val_list[0].
     """
     if isinstance(val_list[0], torch.Tensor):
         return red_func(val_list)
+    elif isinstance(val_list[0], bool):
+        return any(val_list)
     assert isinstance(val_list[0], tuple), "Elements to be reduced can only be"
     "either Tensors or tuples containing Tensors."
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_utils/batching.py]: Yes, _reduce_list expects a list of either tensors or tuples (with all elements corresponding), but we can add support for any iterable if necessary. I think for the line mentioned, it might be fine the way it is, since we intend to reduce all corresponding positioned elements of each tuple (with the assumption that each is the same length), the outer loop just iterates through the positions within a tuple, and the inner call to reduce list obtains all elements of that position.

In your example, the loop should first call _reduce_list with (t,t2,t3) and the second time again with (the second elements) (t, t2, t3). Does that seem correct?

File: captum/attr/_core/noise_tunnel.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/noise_tunnel.py]: I've updated to this approach, the only issues were that InputBaselineXGradient and LayerInputBaselineXGradient were only being called by NoiseTunnel through GradientShap and assumed that inputs were already a tuple, needed to also format inputs in these methods.

File: captum/attr/_utils/batching.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_utils/batching.py]: Oh, I see, maybe we should document it. Because, I was originally thinking that I can expect the elements only within a tuple to have same dimensionality. Is this used here ?https://github.com/pytorch/captum/blob/0997b63641f8495767f522d5a856697d5c5fe653/captum/attr/_utils/gradient.py#L104

File: captum/attr/_core/noise_tunnel.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/noise_tunnel.py]: Great! Thank you!

File: captum/attr/_utils/batching.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_utils/batching.py]: Makes sense, will add a comment to make it clearer! I think here each element in gradient_tensors is a tuple matching the input (so each element of the tuple is not necessarily the same dimensionality), but all the corresponding elements of tuples should match.
